### DISCUSSION TITLE
Operating modes refactor

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#19338] [Feature requests]
From https://github.com/OpenRCT2/OpenRCT2/pull/19305 it was brought up that operating modes could be streamlined with descriptors for operating modes.

A few things that several ride types have in common:

- PoweredLaunch(w+w/o passing station) and PoweredLaunchBlockSection have powered launches
- ContinuousCircuit and PoweredLaunch(w+w/o passing station) can have multiple circuits when the ride has 1 train
- PoweredLaunch w/ passing station and reverse incline launch mode both allow the train to go through the station backwards once
- PoweredLaunchBlockSection and ContinuousCircuitBlockSection use block sections
- Motion simulator ride modes (AvengingAviators, ThrillRiders) are identical in every way. I believe the only reason there's two is for guest attraction - I believe more guests are attracted to two rides with different operating modes than with the same mode.
- 3D Cinema ride modes (MouseTails, StormChasers, SpaceRaiders) differ only by ride length.
- Top spin modes (Beginner, Intermediate, Berserk) differ only by which sub_state the vehicle is set to and the stats the ride gets.

For more discussion about ride modes, see this: https://github.com/OpenRCT2/OpenRCT2/discussions/18967

**Comment by kyphii** [comment]
To reiterate the impetus for this discussion from #19305: Currently RideMode-related logic exists in several locations, frequently in conditional or switch statements which check for a group of ride modes. In order to add new modes, each of these locations need to be touched, which would quickly become unsustainable. 

My suggestion is to define RideMode as a struct rather than a simple enum. Letting the RideModes store some of these shared properties would improve code maintainability in several ways, including eliminating the need to modify the aforementioned locations in code.

The RideMode struct could potentially include:

- Display name (Currently stored in a completely separate source file from other RideMode-related code)
- A set of flags for various properties - IS_BLOCK_SECTIONED, IS_LAUNCHED, IS_FERRIS_WHEEL_ROTATION, etc.
- Corresponding RideMode with blocks enabled/disabled, if one exists
- Required number of stations
- Additional variables for rides with several similar modes - for example, instead of needing a switch statement to derive sub_state from mode, the Motion Sim could simply retrieve the sub_state of its assigned mode

In addition, it's probably possible to refactor a lot of RideMode-dependent vehicle update code into parameterized functions stored in the RideMode struct. This could greatly simplify some of the large conditional chains and switch statements that currently exist in Vehicle.cpp. (Speaking from personal experience, I faced a nearly identical maintenance problem in one of my projects, and this approach resulted in significantly streamlined code). This could also have a slight performance benefit, as ride vehicles will no longer need to check for operating modes they don't have (Instead, they simply call their RideMode's corresponding function, if one exists).

**Reply by spacek531** [reply]
I was imagining patterning it on the ride type descriptor, which is more-or-less what you're describing.
